### ISSUE TITLE
Register doesn't work on Firefox because I need to enter a PIN!?

### ISSUE THREAD

**Issue by vDorst** [kanidm/webauthn-rs#34]
I tested the axtix example on Fedora 32.
Tested key: HyperFIDO Pro Mini, which is a Feitian K8 rebrand/OEM device with a LED and button.

But it is not working on Firefox 82.
I can click register, Firefox asks about to make the keydata more anonymous.
But after that, my key LED doesn't blink.

But when I try Chromium:
Username "Test" and click Register
Chomium asks for a PIN.
After entering the PIN to Unlock
The LED of the key blinks and after pressing the button, registeration is compleed.

Then when I go back to firefox.
Only login works.

So I wonder why Firefox doesn't support a PIN.
Second why does your example needs that PIN? Maybe can we turn that off?
Why does https://demo.yubico.com/webauthn-technical/registration need a pin to register on Firefox?

**Comment by Firstyear** [comment]
Hi there,

It would be good to see the output from `RUST_LOG=webauthn=debug` while running the example. I have a suspicion about what is occurring though. In the example we request the policy `UserVerificationPolicy::Required`. Firefox today I think only supports CTAP1, which is not capable of verification, so the UV Bit will be false (and these bits are part of the collected client data.).

I suspect then that chromium is using CTAP2 because it *does* support user verification (the pin), and that bit can then be true, and you see that the token only works in chromium. 

If you want to check this, in examples/actix/actors.rs change `.generate_challenge_register(&msg.username, Some(UserVerificationPolicy::Required))?;` to `.generate_challenge_register(&msg.username, Some(UserVerificationPolicy::Discouraged))?;`

This in mind, I think the example/demo server needs a bit of a rework so this is a good reminder that I should undertake that work.

**Comment by vDorst** [comment]
```
[2020-10-28T07:29:54Z INFO  actix_web::middleware::logger] [::1]:58596 "GET /auth HTTP/1.1" 200 2225 "-" "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" 0.000087
[2020-10-28T07:30:01Z DEBUG actix::actors] handle ChallengeRegister -> ChallengeRegister { username: "firefox" }
[2020-10-28T07:30:01Z DEBUG actix::actors] complete ChallengeRegister -> CreationChallengeResponse { public_key: PublicKeyCredentialCreationOptions { rp: RelyingParty { name: "localhost", id: "localhost" }, user: User { id: Base64UrlSafeData([102, 105, 114, 101, 102, 111, 120]), name: "firefox", display_name: "firefox" }, challenge: Base64UrlSafeData([255, 167, 252, 38, 104, 78, 186, 85, 186, 22, 182, 239, 109, 116, 154, 233, 150, 182, 186, 167, 85, 49, 228, 178, 38, 221, 109, 180, 107, 182, 2, 17]), pub_key_cred_params: [PubKeyCredParams { type_: "public-key", alg: -7 }, PubKeyCredParams { type_: "public-key", alg: -35 }, PubKeyCredParams { type_: "public-key", alg: -36 }, PubKeyCredParams { type_: "public-key", alg: -257 }, PubKeyCredParams { type_: "public-key", alg: -258 }, PubKeyCredParams { type_: "public-key", alg: -259 }, PubKeyCredParams { type_: "public-key", alg: -37 }, PubKeyCredParams { type_: "public-key", alg: -38 }, PubKeyCredParams { type_: "public-key", alg: -39 }, PubKeyCredParams { type_: "public-key", alg: -8 }], timeout: Some(60000), attestation: Some(Direct), exclude_credentials: None, authenticator_selection: Some(AuthenticatorSelectionCriteria { authenticator_attachment: None, require_resident_key: false, user_verification: Required }), extensions: None } }
[2020-10-28T07:30:01Z INFO  actix_web::middleware::logger] [::1]:58606 "POST /auth/challenge/register/firefox HTTP/1.1" 200 649 "http://localhost:8080/auth" "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" 0.000191
```

With the change `UserVerificationPolicy::Required` to `UserVerificationPolicy::Discouraged`, it works as expected.
I can register and login with Firefox.
```
[2020-10-28T07:32:09Z DEBUG actix] Started logging ...
[2020-10-28T07:32:09Z INFO  actix_server::builder] Starting 4 workers
[2020-10-28T07:32:09Z INFO  actix_server::builder] Starting "actix-web-service-[::1]:8080" service on [::1]:8080
[2020-10-28T07:32:09Z INFO  actix_server::builder] Starting "actix-web-service-127.0.0.1:8080" service on 127.0.0.1:8080
Started http server: localhost
[2020-10-28T07:32:43Z INFO  actix_web::middleware::logger] [::1]:58610 "GET /auth HTTP/1.1" 200 2225 "-" "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" 0.000106
[2020-10-28T07:32:48Z DEBUG actix::actors] handle ChallengeRegister -> ChallengeRegister { username: "firefox" }
[2020-10-28T07:32:48Z DEBUG actix::actors] complete ChallengeRegister -> CreationChallengeResponse { public_key: PublicKeyCredentialCreationOptions { rp: RelyingParty { name: "localhost", id: "localhost" }, user: User { id: Base64UrlSafeData([102, 105, 114, 101, 102, 111, 120]), name: "firefox", display_name: "firefox" }, challenge: Base64UrlSafeData([95, 161, 170, 42, 125, 164, 46, 74, 154, 120, 218, 1, 141, 89, 205, 1, 239, 231, 230, 162, 133, 87, 209, 178, 49, 2, 99, 249, 247, 225, 95, 134]), pub_key_cred_params: [PubKeyCredParams { type_: "public-key", alg: -7 }, PubKeyCredParams { type_: "public-key", alg: -35 }, PubKeyCredParams { type_: "public-key", alg: -36 }, PubKeyCredParams { type_: "public-key", alg: -257 }, PubKeyCredParams { type_: "public-key", alg: -258 }, PubKeyCredParams { type_: "public-key", alg: -259 }, PubKeyCredParams { type_: "public-key", alg: -37 }, PubKeyCredParams { type_: "public-key", alg: -38 }, PubKeyCredParams { type_: "public-key", alg: -39 }, PubKeyCredParams { type_: "public-key", alg: -8 }], timeout: Some(60000), attestation: Some(Direct), exclude_credentials: None, authenticator_selection: Some(AuthenticatorSelectionCriteria { authenticator_attachment: None, require_resident_key: false, user_verification: Discouraged }), extensions: None } }
[2020-10-28T07:32:48Z INFO  actix_web::middleware::logger] [::1]:58614 "POST /auth/challenge/register/firefox HTTP/1.1" 200 652 "http://localhost:8080/auth" "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" 0.000200
[2020-10-28T07:32:52Z DEBUG actix::actors] handle Register -> Register { username: "firefox", reg: RegisterPublicKeyCredential { id: "J1UXwCp7T4mX5NPe-0GeLBwXab_p62BGYxLls6fkmz9KbnFz44ViWtA3np_5BZJTezx8QPh8uG2vDLTqD7sJBez61ZAG8_gYJVdo5sMEzUWyGdNp5xAfnq-UEoRngmZY", raw_id: Base64UrlSafeData([39, 85, 23, 192, 42, 123, 79, 137, 151, 228, 211, 222, 251, 65, 158, 44, 28, 23, 105, 191, 233, 235, 96, 70, 99, 18, 229, 179, 167, 228, 155, 63, 74, 110, 113, 115, 227, 133, 98, 90, 208, 55, 158, 159, 249, 5, 146, 83, 123, 60, 124, 64, 248, 124, 184, 109, 175, 12, 180, 234, 15, 187, 9, 5, 236, 250, 213, 144, 6, 243, 248, 24, 37, 87, 104, 230, 195, 4, 205, 69, 178, 25, 211, 105, 231, 16, 31, 158, 175, 148, 18, 132, 103, 130, 102, 88]), response: AuthenticatorAttestationResponseRaw { attestation_object: Base64UrlSafeData([163, 99, 102, 109, 116, 104, 102, 105, 100, 111, 45, 117, 50, 102, 103, 97, 116, 116, 83, 116, 109, 116, 162, 99, 115, 105, 103, 88, 71, 48, 69, 2, 33, 0, 209, 197, 7, 119, 123, 118, 103, 121, 22, 149, 10, 60, 202, 163, 235, 44, 54, 129, 6, 220, 141, 36, 206, 196, 171, 91, 93, 206, 44, 185, 129, 165, 2, 32, 50, 202, 173, 62, 164, 66, 213, 66, 130, 153, 178, 215, 131, 135, 101, 185, 246, 24, 228, 130, 82, 49, 98, 205, 179, 152, 120, 151, 32, 1, 217, 184, 99, 120, 53, 99, 129, 89, 1, 163, 48, 130, 1, 159, 48, 130, 1, 69, 160, 3, 2, 1, 2, 2, 2, 16, 20, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 2, 48, 58, 49, 11, 48, 9, 6, 3, 85, 4, 6, 19, 2, 67, 65, 49, 18, 48, 16, 6, 3, 85, 4, 10, 12, 9, 72, 89, 80, 69, 82, 83, 69, 67, 85, 49, 23, 48, 21, 6, 3, 85, 4, 3, 12, 14, 72, 89, 80, 69, 82, 70, 73, 68, 79, 32, 48, 50, 48, 48, 48, 32, 23, 13, 49, 56, 48, 49, 48, 49, 48, 48, 48, 48, 48, 48, 90, 24, 15, 50, 48, 51, 50, 49, 50, 51, 49, 50, 51, 53, 57, 53, 57, 90, 48, 61, 49, 11, 48, 9, 6, 3, 85, 4, 6, 19, 2, 67, 65, 49, 18, 48, 16, 6, 3, 85, 4, 10, 12, 9, 72, 89, 80, 69, 82, 83, 69, 67, 85, 49, 26, 48, 24, 6, 3, 85, 4, 3, 12, 17, 72, 89, 80, 69, 82, 70, 73, 68, 79, 32, 48, 50, 48, 48, 32, 49, 56, 48, 89, 48, 19, 6, 7, 42, 134, 72, 206, 61, 2, 1, 6, 8, 42, 134, 72, 206, 61, 3, 1, 7, 3, 66, 0, 4, 13, 0, 68, 45, 160, 254, 179, 115, 77, 17, 108, 254, 178, 173, 91, 121, 16, 106, 230, 99, 66, 179, 85, 191, 202, 98, 169, 3, 187, 140, 152, 68, 193, 240, 97, 1, 90, 249, 211, 157, 2, 170, 92, 186, 226, 242, 223, 189, 34, 104, 19, 193, 47, 223, 22, 222, 205, 70, 164, 55, 151, 124, 253, 78, 163, 54, 48, 52, 48, 29, 6, 3, 85, 29, 14, 4, 22, 4, 20, 95, 188, 75, 167, 83, 5, 33, 135, 170, 179, 199, 65, 209, 249, 236, 111, 179, 196, 216, 117, 48, 19, 6, 11, 43, 6, 1, 4, 1, 130, 229, 28, 2, 1, 1, 4, 4, 3, 2, 5, 32, 48, 10, 6, 8, 42, 134, 72, 206, 61, 4, 3, 2, 3, 72, 0, 48, 69, 2, 33, 0, 251, 82, 203, 28, 91, 195, 115, 122, 180, 200, 75, 156, 12, 118, 152, 223, 57, 0, 226, 195, 130, 184, 193, 66, 13, 139, 167, 95, 174, 20, 241, 42, 2, 32, 54, 18, 202, 174, 79, 242, 159, 217, 94, 99, 250, 241, 157, 109, 52, 147, 207, 229, 7, 101, 86, 206, 5, 40, 90, 57, 209, 24, 192, 230, 188, 116, 104, 97, 117, 116, 104, 68, 97, 116, 97, 88, 228, 73, 150, 13, 229, 136, 14, 140, 104, 116, 52, 23, 15, 100, 118, 96, 91, 143, 228, 174, 185, 162, 134, 50, 199, 153, 92, 243, 186, 131, 29, 151, 99, 65, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 96, 39, 85, 23, 192, 42, 123, 79, 137, 151, 228, 211, 222, 251, 65, 158, 44, 28, 23, 105, 191, 233, 235, 96, 70, 99, 18, 229, 179, 167, 228, 155, 63, 74, 110, 113, 115, 227, 133, 98, 90, 208, 55, 158, 159, 249, 5, 146, 83, 123, 60, 124, 64, 248, 124, 184, 109, 175, 12, 180, 234, 15, 187, 9, 5, 236, 250, 213, 144, 6, 243, 248, 24, 37, 87, 104, 230, 195, 4, 205, 69, 178, 25, 211, 105, 231, 16, 31, 158, 175, 148, 18, 132, 103, 130, 102, 88, 165, 1, 2, 3, 38, 32, 1, 33, 88, 32, 121, 226, 79, 109, 168, 150, 255, 214, 19, 209, 227, 223, 16, 59, 62, 245, 221, 157, 253, 98, 127, 42, 133, 35, 179, 141, 45, 206, 146, 172, 48, 48, 34, 88, 32, 91, 98, 160, 125, 190, 42, 94, 218, 113, 102, 193, 76, 9, 41, 122, 126, 32, 158, 96, 86, 66, 209, 85, 186, 214, 138, 26, 5, 54, 187, 153, 111]), client_data_json: Base64UrlSafeData([123, 34, 99, 104, 97, 108, 108, 101, 110, 103, 101, 34, 58, 34, 88, 54, 71, 113, 75, 110, 50, 107, 76, 107, 113, 97, 101, 78, 111, 66, 106, 86, 110, 78, 65, 101, 95, 110, 53, 113, 75, 70, 86, 57, 71, 121, 77, 81, 74, 106, 45, 102, 102, 104, 88, 52, 89, 34, 44, 34, 99, 108, 105, 101, 110, 116, 69, 120, 116, 101, 110, 115, 105, 111, 110, 115, 34, 58, 123, 125, 44, 34, 104, 97, 115, 104, 65, 108, 103, 111, 114, 105, 116, 104, 109, 34, 58, 34, 83, 72, 65, 45, 50, 53, 54, 34, 44, 34, 111, 114, 105, 103, 105, 110, 34, 58, 34, 104, 116, 116, 112, 58, 47, 47, 108, 111, 99, 97, 108, 104, 111, 115, 116, 58, 56, 48, 56, 48, 34, 44, 34, 116, 121, 112, 101, 34, 58, 34, 119, 101, 98, 97, 117, 116, 104, 110, 46, 99, 114, 101, 97, 116, 101, 34, 125]) }, type_: "public-key" } }
[2020-10-28T07:32:52Z DEBUG webauthn_rs] attestation is: FIDOU2F
[2020-10-28T07:32:52Z DEBUG webauthn_rs] policy_verify_trust -> Basic(Credential { cred_id: [39, 85, 23, 192, 42, 123, 79, 137, 151, 228, 211, 222, 251, 65, 158, 44, 28, 23, 105, 191, 233, 235, 96, 70, 99, 18, 229, 179, 167, 228, 155, 63, 74, 110, 113, 115, 227, 133, 98, 90, 208, 55, 158, 159, 249, 5, 146, 83, 123, 60, 124, 64, 248, 124, 184, 109, 175, 12, 180, 234, 15, 187, 9, 5, 236, 250, 213, 144, 6, 243, 248, 24, 37, 87, 104, 230, 195, 4, 205, 69, 178, 25, 211, 105, 231, 16, 31, 158, 175, 148, 18, 132, 103, 130, 102, 88], cred: COSEKey { type_: ECDSA_SHA256, key: EC_EC2(COSEEC2Key { curve: SECP256R1, x: [121, 226, 79, 109, 168, 150, 255, 214, 19, 209, 227, 223, 16, 59, 62, 245, 221, 157, 253, 98, 127, 42, 133, 35, 179, 141, 45, 206, 146, 172, 48, 48], y: [91, 98, 160, 125, 190, 42, 94, 218, 113, 102, 193, 76, 9, 41, 122, 126, 32, 158, 96, 86, 66, 209, 85, 186, 214, 138, 26, 5, 54, 187, 153, 111] }) }, counter: 0 }, X509PublicKey)
[2020-10-28T07:32:52Z DEBUG actix::actors] {[102, 105, 114, 101, 102, 111, 120]: {[39, 85, 23, 192, 42, 123, 79, 137, 151, 228, 211, 222, 251, 65, 158, 44, 28, 23, 105, 191, 233, 235, 96, 70, 99, 18, 229, 179, 167, 228, 155, 63, 74, 110, 113, 115, 227, 133, 98, 90, 208, 55, 158, 159, 249, 5, 146, 83, 123, 60, 124, 64, 248, 124, 184, 109, 175, 12, 180, 234, 15, 187, 9, 5, 236, 250, 213, 144, 6, 243, 248, 24, 37, 87, 104, 230, 195, 4, 205, 69, 178, 25, 211, 105, 231, 16, 31, 158, 175, 148, 18, 132, 103, 130, 102, 88]: Credential { cred_id: [39, 85, 23, 192, 42, 123, 79, 137, 151, 228, 211, 222, 251, 65, 158, 44, 28, 23, 105, 191, 233, 235, 96, 70, 99, 18, 229, 179, 167, 228, 155, 63, 74, 110, 113, 115, 227, 133, 98, 90, 208, 55, 158, 159, 249, 5, 146, 83, 123, 60, 124, 64, 248, 124, 184, 109, 175, 12, 180, 234, 15, 187, 9, 5, 236, 250, 213, 144, 6, 243, 248, 24, 37, 87, 104, 230, 195, 4, 205, 69, 178, 25, 211, 105, 231, 16, 31, 158, 175, 148, 18, 132, 103, 130, 102, 88], cred: COSEKey { type_: ECDSA_SHA256, key: EC_EC2(COSEEC2Key { curve: SECP256R1, x: [121, 226, 79, 109, 168, 150, 255, 214, 19, 209, 227, 223, 16, 59, 62, 245, 221, 157, 253, 98, 127, 42, 133, 35, 179, 141, 45, 206, 146, 172, 48, 48], y: [91, 98, 160, 125, 190, 42, 94, 218, 113, 102, 193, 76, 9, 41, 122, 126, 32, 158, 96, 86, 66, 209, 85, 186, 214, 138, 26, 5, 54, 187, 153, 111] }) }, counter: 0 }}}
[2020-10-28T07:32:52Z DEBUG actix::actors] complete Register -> Ok(())
[2020-10-28T07:32:52Z INFO  actix_web::middleware::logger] [::1]:58614 "POST /auth/register/firefox HTTP/1.1" 200 4 "http://localhost:8080/auth" "Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:82.0) Gecko/20100101 Firefox/82.0" 0.002273
```

If you need more information/testing, please let me know.

**Comment by Firstyear** [comment]
Yes, it may be that #32 is related in this case. I've been writing webauthn-authenticator-rs and that's what gave me the insight about ctap1/2 recently.

So now I'm curious about what happens with UserVerificationPolicy::Preferred. I think in Preferred authenticator-rs in firefox will continue as though "discouraged", but then if you use the same token in chrome it will use it as though "preferred". I'm wondering if that works .... I will do some of my own testing of this later, but if you want to test and confirm the behaviour for me too that would be great.

**Comment by Firstyear** [comment]
Okay, I have confirmed this behaviour. If you have firefox and a yubikey and UVP::Required, registration can *not* proceed. However, if you register the device on chrome/edge with Required, then move to Firefox, because generate_challenge_authenticate() takes a policy of "Option" defaulting to "Preferred" then you *can* authenticate on FF because CTAP1 is able to proceed without the verification.

So I think in this situation, this creates an ambiguity, where the credential can have it's verification setting altered between the initial registration and then the authentication. This exposes a risk as it means a verified credential can then potentially be used *without* verification at a later point.

I think to resolve this, the correct action is to embed the verified bit into the generated credential during registration, and rather that needing a policy on generate_challenge_authenticate, the verification bit of the credential determines if the subsequent verification is required. This means the policy for the credential is set at registration time, rather than per-authentication, which gives a better and consistent experience.

Additionally, I think the example needs UVP::Discouraged to make this a bit smoother too.

**Comment by Firstyear** [comment]
I think this may actually be a defect in the webauthn standard, because given a list of credentials for challenge authenticate, the UVP is for the entire authentication, not per-credential. This means you can have a list of credentials where some may be verified, and some may *not* be verified, and then you have to give a policy for the whole operation where either Required creds become "un-verified" or creds signed up without verification capability suddenly attempt to request it. I'm going to need to think about how to handle and express this.

**Comment by Firstyear** [comment]
I have opened https://github.com/w3c/webauthn/issues/1510 about this.

**Comment by Firstyear** [comment]
Okay, so after some comments there, I think that I'm going to make some changes to the library based around UV policies and how they are exposed in this server. I have an inprogress branch to resolve this, but there are quite a few changes at the same time, so it's likely a major version/breaking change.

**Comment by vDorst** [comment]
For me that is no issue.
I still exploring how everything works and fits together.

**Comment by Firstyear** [comment]
Thanks @vDorst, I'll let you know when it's published.

**Comment by Firstyear** [comment]
Published `v0.3.0-alpha.1` which changes this experimentally. Very open to feedback on it.
